src/scroll-timeline-css-parser.js

```diff
@@ -0,0 +1,360 @@
+  }
+
+  parseQualifiedRule(p) {
+    const startIndex = p.index;
+    const selector = this.parseSelector(p).trim();
+    if (!selector) return;
+    const block = this.eatBlock(p);
+    const endIndex = p.index;
+    return {
+      selector,
+      block,
+      startIndex,
+      endIndex,
+    };
+  }
+
+  removeEnclosingDoubleQuotes(s) {
+    let startIndex = s[0] == '"' ? 1 : 0;
+    let endIndex = s[s.length - 1] == '"' ? s.length - 1 : s.length;
+    return s.substring(startIndex, endIndex);
+  }
+
+  assertString(p, s) {
+    if (p.sheetSrc.substr(p.index, s.length) != s) {
+      throw this.parseError(p, `Did not find expected sequence ${s}`);
+    }
+    p.index += s.length;
+  }
+
+  replacePart(start, end, replacement, p) {
+    p.sheetSrc = p.sheetSrc.slice(0, start) + replacement + p.sheetSrc.slice(end);
+    // If we are pointing past the end of the affected section, we need to
+    // recalculate the string pointer. Pointing to something inside the section
+    // that’s being replaced is undefined behavior. Sue me.
+    if (p.index >= end) {
+      const delta = p.index - end;
+      p.index = start + replacement.length + delta;
+    }
+  }
+
+  eatComment(p) {
+    this.assertString(p, "/*");
+    this.eatUntil("*/", p);
+    this.assertString(p, "*/");
+  }
+
+  eatBlock(p) {
+    const startIndex = p.index;
+    this.assertString(p, "{");
+    let level = 1;
+    while (level != 0) {
+      if (p.sheetSrc[p.index] === "{") {
+        level++;
+      } else if (p.sheetSrc[p.index] === "}") {
+        level--;
+      }
+      this.advance(p);
+    }
+    const endIndex = p.index;
+    const contents = p.sheetSrc.slice(startIndex, endIndex);
+
+    return { startIndex, endIndex, contents };
+  }
+
+  advance(p) {
+    p.index++;
+    if (p.index > p.sheetSrc.length) {
+      throw this.parseError(p, "Advanced beyond the end");
+    }
+  }
+
+  eatUntil(s, p) {
+    const startIndex = p.index;
+    while (!this.lookAhead(s, p)) {
+      this.advance(p);
+    }
+    return p.sheetSrc.slice(startIndex, p.index);
+  }
+
+  parseSelector(p) {
+    let startIndex = p.index;
+    this.eatUntil("{", p);
+    if (startIndex === p.index) {
+      throw Error("Empty selector");
+    }
+
+    return p.sheetSrc.slice(startIndex, p.index);
+  }
+
+  eatWhitespace(p) {
+    // Start matching at the current position in the sheet src
+    RegexMatcher.WHITE_SPACE.lastIndex = p.index;
+    const match = RegexMatcher.WHITE_SPACE.exec(p.sheetSrc);
+    if (match) {
+      p.index += match[0].length;
+    }
+  }
+
+  lookAhead(s, p) {
+    return p.sheetSrc.substr(p.index, s.length) == s;
+  }
+
+  peek(p) {
+    return p.sheetSrc[p.index];
+  }
+}
+
+const VALID_SCROLL_OFFSET_SUFFIXES = [
+  // Relative lengths.
+  'em',
+  'ex',
+  'ch',
+  'rem',
+  'vw',
+  'vh',
+  'vmin',
+  'vmax',
+  // Absolute lengths.
+  'cm',
+  'mm',
+  'q',
+  'in',
+  'pc',
+  'pt',
+  'px',
+  // Percentage.
+  '%',
+];
+
+export const RegexMatcher = {
```

Comment:
Suggest moving constants near the top of the file.  Though we only need these constants defined before we use them, it is more difficult  to read the code if needing to frequently look ahead to see the definitions.